drivers/led/led_gpio.c

```diff
@@ -0,0 +1,171 @@
+}
+
+static inline const struct led_gpio_cfg *get_dev_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+static struct led_gpio_data *led_parameters_get(struct device *dev, u32_t led)
+{
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		if (data[i].cfg->id == led) {
+			return &data[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+{
+	int err;
+	u32_t pin_value;
+
+	data->state = state;
+
+	if (data->cfg->gpio.polarity & GPIO_INT_ACTIVE_HIGH) {
+		pin_value = (state ? 1 : 0);
+	} else {
+		pin_value = (state ? 0 : 1);
+	}
+
+	err = gpio_pin_write(data->gpio_dev, data->cfg->gpio.pin,
+			     pin_value);
+	if (err) {
+		LOG_DBG("%s turn %s error %d",
+			data->cfg->label,
+			state ? "on" : "off",
+			err);
+	} else {
+		LOG_DBG("%s turn %s", data->cfg->label,
+			state ? "on" : "off");
+	}
+
+	return err;
+}
+
+static int gpio_led_set(struct device *dev, u32_t index, bool state)
+{
+	return gpio_led_set_no_delay(data, state);
+}
+
+static int gpio_led_on(struct device *dev, u32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, true);
+}
+
+static int gpio_led_off(struct device *dev, u32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, false);
+}
+
+static const struct led_driver_api led_api = {
+	.on = gpio_led_on,
+	.off = gpio_led_off,
+};
+
+static int gpio_led_init(struct device *dev)
+{
+	int err;
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+	u32_t pin_value;
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		data[i].gpio_dev = device_get_binding(cfg->led[i].gpio.port);
+		if (!data[i].gpio_dev) {
+			LOG_DBG("Failed to get GPIO device for port %s pin %d",
+				cfg->led[i].gpio.port, cfg->led[i].gpio.pin);
+			return -EINVAL;
+		}
+
+		/* Set all leds pin as output */
+		err = gpio_pin_configure(data[i].gpio_dev, cfg->led[i].gpio.pin,
+					 GPIO_DIR_OUT);
+		if (err) {
+			LOG_DBG("Failed to set GPIO port %s pin %d as output",
+				cfg->led[i].gpio.port, cfg->led[i].gpio.pin);
+			return err;
+		}
+
+		if (cfg->led[i].gpio.polarity & GPIO_INT_ACTIVE_HIGH) {
+			pin_value = 0;
+		} else {
+			pin_value = 1;
+		}
+
+		err = gpio_pin_write(data[i].gpio_dev, cfg->led[i].gpio.pin,
+				     pin_value);
+		if (err) {
+			LOG_DBG("Failed to turn off %s", cfg->led[i].label);
+		}
+
+		data[i].cfg = &cfg->led[i];
+	}
+
+	return 0;
+}
+
+static const struct led_dts_cfg led_dts[] = DT_INST_0_GPIO_LEDS_INIT;
+static struct led_gpio_data led_data[ARRAY_SIZE(led_dts)];
+static const struct led_gpio_cfg led_cfg = {
+	.led_cnt = ARRAY_SIZE(led_dts),
+	.led = led_dts
+};
+
+DEVICE_AND_API_INIT(gpio_led, DT_INST_0_GPIO_LEDS_DRIVER,
```

Comment:
Maybe it is possible to support all the gpio-led instances ? With this new gpio-led controller notion (with LEDs numbering), someone may want to group the GPIO LEDs into dedicated nodes.